File: TextosJapones/src/main/java/org/jisho/textosJapones/controller/mangas/MangasVincularCelulaDuplaController.java
package org.jisho.textosJapones.controller.mangas;

import javafx.css.PseudoClass;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.image.ImageView;
import javafx.scene.input.*;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.scene.text.Text;
import org.jisho.textosJapones.components.listener.VinculoListener;
import org.jisho.textosJapones.model.entities.VinculoPagina;
import org.jisho.textosJapones.model.enums.Pagina;
import org.jisho.textosJapones.util.Util;

import java.net.URL;
import java.util.ResourceBundle;

public class MangasVincularCelulaDuplaController implements Initializable {

	final PseudoClass ON_DRAG_SELECIONADO = PseudoClass.getPseudoClass("drag-selecionado");

	@FXML
	public HBox root;

	@FXML
	public AnchorPane esquerdaContainer;

	@FXML
	public ImageView esquerdaImagem;

	@FXML
	public Text esquerdaNumero;

	@FXML
	public VBox esquerdaNome;

	@FXML
	public Text esquerdaNomePagina;

	@FXML
	public AnchorPane direitaContainer;

	@FXML
	public ImageView direitaImagem;

	@FXML
	public Text direitaNumero;

	@FXML
	public VBox direitaNome;

	@FXML
	public Text direitaNomePagina;

	private VinculoListener listener;

	public void setDados(VinculoPagina vinculo) {
		root.setOnDragEntered(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				if (vinculo.getVinculadoEsquerdaPagina() != VinculoPagina.PAGINA_VAZIA) {
					direitaContainer.setVisible(true);
					direitaContainer.setPrefWidth(esquerdaContainer.getWidth());
				}

			}
		});

		root.setOnDragExited(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				if (!vinculo.isImagemDupla && vinculo.getVinculadoDireitaPagina() == VinculoPagina.PAGINA_VAZIA)
					direitaContainer.setVisible(false);

			}
		});

		root.setOnDragDone(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				if (!vinculo.isImagemDupla && vinculo.getVinculadoDireitaPagina() == VinculoPagina.PAGINA_VAZIA)
					direitaContainer.setVisible(false);
			}
		});
		
		if (vinculo != null) {
			configuraDrop(esquerdaContainer, vinculo, Pagina.VINCULADO_ESQUERDA);
			if (vinculo.getVinculadoEsquerdaPagina() == VinculoPagina.PAGINA_VAZIA) {
				esquerdaContainer.setOnDragDetected(null);
				esquerdaImagem.setImage(vinculo.getImagemVinculadoEsquerda());
				esquerdaNumero.setText("");
				esquerdaNomePagina.setText("");
				esquerdaNome.setVisible(false);
			} else {
				esquerdaContainer.setOnDragDetected(new EventHandler<MouseEvent>() {
					public void handle(MouseEvent event) {
						listener.onDragStart();
						Dragboard db = esquerdaContainer.startDragAndDrop(TransferMode.ANY);

						db.setDragView(Util.criaSnapshot(esquerdaContainer), event.getX(), event.getY());

						vinculo.onDragOrigem = Pagina.VINCULADO_ESQUERDA;
						ClipboardContent content = new ClipboardContent();
						content.put(Util.VINCULO_ITEM_FORMAT, vinculo);
						content.put(Util.NUMERO_PAGINA_ITEM_FORMAT, vinculo.getVinculadoEsquerdaPagina().toString());
						content.putString(vinculo.getVinculadoEsquerdaNomePagina());
						db.setContent(content);

						event.consume();
					}
				});

				esquerdaContainer.setOnDragDone(new EventHandler<DragEvent>() {
					public void handle(DragEvent event) {
						listener.onDragEnd();
						direitaContainer.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false);
					}
				});

				esquerdaImagem.setImage(vinculo.getImagemVinculadoEsquerda());
				esquerdaNumero.setText(vinculo.getVinculadoEsquerdaPagina().toString());
				esquerdaNomePagina.setText(vinculo.getVinculadoEsquerdaNomePagina());
				esquerdaNome.setVisible(true);
			}

			configuraDrop(direitaContainer, vinculo, Pagina.VINCULADO_DIREITA);

			if (vinculo.getVinculadoDireitaPagina() == VinculoPagina.PAGINA_VAZIA) {
				direitaContainer.setOnDragDetected(null);
				direitaContainer.setVisible(false);
				direitaImagem.setImage(vinculo.getImagemVinculadoDireita());
				direitaNumero.setText("");
				direitaNomePagina.setText("");
			} else {
				direitaContainer.setOnDragDetected(new EventHandler<MouseEvent>() {
					public void handle(MouseEvent event) {
						listener.onDragStart();

						Dragboard db = esquerdaContainer.startDragAndDrop(TransferMode.ANY);

						vinculo.onDragOrigem = Pagina.VINCULADO_DIREITA;
						ClipboardContent content = new ClipboardContent();
						content.put(Util.VINCULO_ITEM_FORMAT, vinculo);
						content.put(Util.NUMERO_PAGINA_ITEM_FORMAT, vinculo.getVinculadoEsquerdaPagina().toString());
						content.putString(vinculo.getVinculadoEsquerdaNomePagina());
						db.setContent(content);

						event.consume();
					}
				});

				direitaContainer.setOnDragDone(new EventHandler<DragEvent>() {
					public void handle(DragEvent event) {
						listener.onDragEnd();
						direitaContainer.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false);
					}
				});

				direitaContainer.setVisible(vinculo.getImagemVinculadoDireita() != null);
				direitaImagem.setImage(vinculo.getImagemVinculadoDireita());
				direitaNumero.setText(vinculo.getVinculadoDireitaPagina().toString());
				direitaNomePagina.setText(vinculo.getVinculadoDireitaNomePagina());
			}
		} else {
			configuraDrop(esquerdaContainer, vinculo, Pagina.VINCULADO_DIREITA);
			esquerdaImagem.setImage(null);
			esquerdaNumero.setText("");
			esquerdaNomePagina.setText("");
			esquerdaNome.setVisible(false);

			direitaContainer.setVisible(false);
			direitaImagem.setImage(null);
			direitaNumero.setText("");
			direitaNomePagina.setText("");
			limpaDrop(direitaContainer);
		}
	}

	private void limpaDrop(AnchorPane pane) {
		pane.setOnDragEntered(null);
		pane.setOnDragExited(null);
	}

	private void configuraDrop(AnchorPane pane, VinculoPagina destino, Pagina tipo) {
		pane.setOnDragOver(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				if (event.getGestureSource() != pane && event.getDragboard().hasContent(Util.VINCULO_ITEM_FORMAT))
					event.acceptTransferModes(TransferMode.COPY_OR_MOVE);

				event.consume();
			}
		});

		pane.setOnDragEntered(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				if (event.getGestureSource() != pane)
					pane.pseudoClassStateChanged(ON_DRAG_SELECIONADO, true);

				event.consume();
			}
		});

		pane.setOnDragExited(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				pane.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false);

				event.consume();
			}
		});

		pane.setOnDragDropped(new EventHandler<DragEvent>() {
			public void handle(DragEvent event) {
				Dragboard db = event.getDragboard();
				boolean success = false;
				if (db.hasContent(Util.VINCULO_ITEM_FORMAT)) {
					VinculoPagina vinculo = (VinculoPagina) db.getContent(Util.VINCULO_ITEM_FORMAT);

					listener.onDrop(vinculo.onDragOrigem, vinculo, tipo, destino);
					success = true;
				}
				event.setDropCompleted(success);
				event.consume();
			}
		});
	}

	public void setListener(VinculoListener listener) {
		this.listener = listener;
	}

	@Override
	public void initialize(URL location, ResourceBundle resources) {

		esquerdaImagem.fitWidthProperty().bind(esquerdaContainer.widthProperty());
		esquerdaImagem.fitHeightProperty().bind(esquerdaContainer.heightProperty());
		esquerdaImagem.setPreserveRatio(true);
		esquerdaImagem.fitWidthProperty().bind(esquerdaNome.widthProperty());

		direitaImagem.fitWidthProperty().bind(direitaContainer.widthProperty());
		direitaImagem.fitHeightProperty().bind(direitaContainer.heightProperty());
		direitaImagem.setPreserveRatio(true);
		direitaImagem.fitWidthProperty().bind(direitaNome.widthProperty());

		direitaContainer.managedProperty().bind(direitaContainer.visibleProperty());

	}

	public static URL getFxmlLocate() {
		return MangasVincularCelulaDuplaController.class.getResource("/view/mangas/MangaVincularCelulaDupla.fxml");
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/controller/mangas/MangasVincularCelulaDuplaController.kt
package br.com.fenix.processatexto.controller.mangas

import br.com.fenix.processatexto.components.listener.VinculoListener
import br.com.fenix.processatexto.model.entities.processatexto.VinculoPagina
import br.com.fenix.processatexto.model.enums.Pagina
import br.com.fenix.processatexto.util.Utils
import javafx.css.PseudoClass
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.image.ImageView
import javafx.scene.input.ClipboardContent
import javafx.scene.input.Dragboard
import javafx.scene.input.TransferMode
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.scene.text.Text
import java.net.URL
import java.util.*


class MangasVincularCelulaDuplaController : Initializable {

    val ON_DRAG_SELECIONADO: PseudoClass = PseudoClass.getPseudoClass("drag-selecionado")

    @FXML
    lateinit var root: HBox

    @FXML
    lateinit var esquerdaContainer: AnchorPane

    @FXML
    lateinit var esquerdaImagem: ImageView

    @FXML
    lateinit var esquerdaNumero: Text

    @FXML
    lateinit var esquerdaNome: VBox

    @FXML
    lateinit var esquerdaNomePagina: Text

    @FXML
    lateinit var direitaContainer: AnchorPane

    @FXML
    lateinit var direitaImagem: ImageView

    @FXML
    lateinit var direitaNumero: Text

    @FXML
    lateinit var direitaNome: VBox

    @FXML
    lateinit var direitaNomePagina: Text

    private var listener: VinculoListener? = null
    fun setDados(vinculo: VinculoPagina) {
        root.setOnDragEntered {
            if (vinculo.vinculadoEsquerdaPagina !== VinculoPagina.PAGINA_VAZIA) {
                direitaContainer.isVisible = true
                direitaContainer.prefWidth = esquerdaContainer.width
            }
        }
        root.setOnDragExited { if (!vinculo.isImagemDupla && vinculo.vinculadoDireitaPagina === VinculoPagina.PAGINA_VAZIA) direitaContainer.isVisible = false }
        root.setOnDragDone { if (!vinculo.isImagemDupla && vinculo.vinculadoDireitaPagina === VinculoPagina.PAGINA_VAZIA) direitaContainer.isVisible = false }
        if (vinculo != null) {
            configuraDrop(esquerdaContainer, vinculo, Pagina.VINCULADO_ESQUERDA)
            if (vinculo.vinculadoEsquerdaPagina === VinculoPagina.PAGINA_VAZIA) {
                esquerdaContainer.onDragDetected = null
                esquerdaImagem.image = vinculo.imagemVinculadoEsquerda
                esquerdaNumero.text = ""
                esquerdaNomePagina.text = ""
                esquerdaNome.isVisible = false
            } else {
                esquerdaContainer.setOnDragDetected { event ->
                    listener!!.onDragStart()
                    val db: Dragboard = esquerdaContainer.startDragAndDrop(TransferMode.COPY, TransferMode.MOVE, TransferMode.LINK)
                    db.setDragView(Utils.criaSnapshot(esquerdaContainer), event.x, event.y)
                    vinculo.onDragOrigem = Pagina.VINCULADO_ESQUERDA
                    val content = ClipboardContent()
                    content[Utils.VINCULO_ITEM_FORMAT] = vinculo
                    content[Utils.NUMERO_PAGINA_ITEM_FORMAT] = vinculo.vinculadoEsquerdaPagina.toString()
                    content.putString(vinculo.vinculadoEsquerdaNomePagina)
                    db.setContent(content)
                    event.consume()
                }
                esquerdaContainer.setOnDragDone {
                    listener!!.onDragEnd()
                    direitaContainer.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false)
                }
                esquerdaImagem.image = vinculo.imagemVinculadoEsquerda
                esquerdaNumero.text = vinculo.vinculadoEsquerdaPagina.toString()
                esquerdaNomePagina.text = vinculo.vinculadoEsquerdaNomePagina
                esquerdaNome.isVisible = true
            }
            configuraDrop(direitaContainer, vinculo, Pagina.VINCULADO_DIREITA)
            if (vinculo.vinculadoDireitaPagina === VinculoPagina.PAGINA_VAZIA) {
                direitaContainer.onDragDetected = null
                direitaContainer.isVisible = false
                direitaImagem.image = vinculo.imagemVinculadoDireita
                direitaNumero.text = ""
                direitaNomePagina.text = ""
            } else {
                direitaContainer.setOnDragDetected { event ->
                    listener!!.onDragStart()
                    val db: Dragboard = esquerdaContainer.startDragAndDrop(TransferMode.COPY, TransferMode.MOVE, TransferMode.LINK)
                    vinculo.onDragOrigem = Pagina.VINCULADO_DIREITA
                    val content = ClipboardContent()
                    content[Utils.VINCULO_ITEM_FORMAT] = vinculo
                    content[Utils.NUMERO_PAGINA_ITEM_FORMAT] = vinculo.vinculadoEsquerdaPagina.toString()
                    content.putString(vinculo.vinculadoEsquerdaNomePagina)
                    db.setContent(content)
                    event.consume()
                }
                direitaContainer.setOnDragDone {
                    listener!!.onDragEnd()
                    direitaContainer.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false)
                }
                direitaContainer.isVisible = vinculo.imagemVinculadoDireita != null
                direitaImagem.image = vinculo.imagemVinculadoDireita
                direitaNumero.text = vinculo.vinculadoDireitaPagina.toString()
                direitaNomePagina.text = vinculo.vinculadoDireitaNomePagina
            }
        } else {
            configuraDrop(esquerdaContainer, vinculo, Pagina.VINCULADO_DIREITA)
            esquerdaImagem.image = null
            esquerdaNumero.text = ""
            esquerdaNomePagina.text = ""
            esquerdaNome.isVisible = false
            direitaContainer.isVisible = false
            direitaImagem.image = null
            direitaNumero.text = ""
            direitaNomePagina.text = ""
            limpaDrop(direitaContainer)
        }
    }

    private fun limpaDrop(pane: AnchorPane) {
        pane.onDragEntered = null
        pane.onDragExited = null
    }

    private fun configuraDrop(pane: AnchorPane, destino: VinculoPagina?, tipo: Pagina) {
        pane.setOnDragOver { event ->
            if (event.gestureSource !== pane && event.dragboard.hasContent(Utils.VINCULO_ITEM_FORMAT))
                event.acceptTransferModes(TransferMode.COPY, TransferMode.MOVE)
            event.consume()
        }
        pane.setOnDragEntered { event ->
            if (event.gestureSource !== pane)
                pane.pseudoClassStateChanged(ON_DRAG_SELECIONADO, true)
            event.consume()
        }
        pane.setOnDragExited { event ->
            pane.pseudoClassStateChanged(ON_DRAG_SELECIONADO, false)
            event.consume()
        }
        pane.setOnDragDropped { event ->
            val db: Dragboard = event.dragboard
            var success = false
            if (db.hasContent(Utils.VINCULO_ITEM_FORMAT)) {
                val vinculo: VinculoPagina = db.getContent(Utils.VINCULO_ITEM_FORMAT) as VinculoPagina
                listener!!.onDrop(vinculo.onDragOrigem!!, vinculo, tipo, destino!!)
                success = true
            }
            event.isDropCompleted = success
            event.consume()
        }
    }

    fun setListener(listener: VinculoListener?) {
        this.listener = listener
    }

    @Override
    override fun initialize(location: URL, resources: ResourceBundle) {
        esquerdaImagem.fitWidthProperty().bind(esquerdaContainer.widthProperty())
        esquerdaImagem.fitHeightProperty().bind(esquerdaContainer.heightProperty())
        esquerdaImagem.isPreserveRatio = true
        esquerdaImagem.fitWidthProperty().bind(esquerdaNome.widthProperty())
        direitaImagem.fitWidthProperty().bind(direitaContainer.widthProperty())
        direitaImagem.fitHeightProperty().bind(direitaContainer.heightProperty())
        direitaImagem.isPreserveRatio = true
        direitaImagem.fitWidthProperty().bind(direitaNome.widthProperty())
        direitaContainer.managedProperty().bind(direitaContainer.visibleProperty())
    }

    companion object {
        val fxmlLocate: URL get() = MangasVincularCelulaDuplaController::class.java.getResource("/view/mangas/MangaVincularCelulaDupla.fxml") as URL
    }
}